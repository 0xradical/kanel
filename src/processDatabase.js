import chalk from 'chalk';
import knex from 'knex';
import rmfr from 'rmfr';
import fs from 'fs';
import { extractSchema } from 'extract-pg-schema';
import defaultTypeMap from './defaultTypeMap';
import { logger } from './logger';
import processSchema from './processSchema';

/**
 * @param {import('./Config').default} config
 */
const processDatabase = async ({
  connection,
  sourceCasing = null,
  typeCasing = null,
  propertyCasing = null,
  filenameCasing = null,
  preDeleteModelFolder = false,
  customTypeMap = {},
  schemas,
}) => {
  const typeMap = { ...defaultTypeMap, ...customTypeMap };
  const casings = { sourceCasing, typeCasing, propertyCasing, filenameCasing };

  logger.log(
    `Connecting to ${chalk.greenBright(connection.database)} on ${
      connection.host
    }`
  );
  const knexConfig = {
    client: 'pg',
    connection,
  };

  for (const schema of schemas) {
    if (preDeleteModelFolder) {
      logger.log(` - Clearing old files in ${schema.modelFolder}`);
      await rmfr(schema.modelFolder, { glob: true });
    }
    if (!fs.existsSync(schema.modelFolder)) {
      fs.mkdirSync(schema.modelFolder);
    }

    const db = knex(knexConfig);
    const extractedSchemaObject = await extractSchema(schema.name, db);

    await processSchema(schema, extractedSchemaObject, typeMap, casings);
  }
};

export default processDatabase;
