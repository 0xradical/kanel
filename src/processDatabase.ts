import chalk from 'chalk';
import { extractSchema } from 'extract-pg-schema';
import fs from 'fs';
import path from 'path';
import { identity, indexBy, isEmpty, map } from 'ramda';
import rmfr from 'rmfr';

import Config, { Hook, nameIdentity, SchemaConfig } from './Config';
import defaultTypeMap from './defaultTypeMap';
import { logger } from './logger';
import processSchema from './processSchema';

const labelAsGenerated: Hook<unknown> = (lines) => [
  '// @generated',
  "// Automatically generated. Don't change this file manually.",
  '',
  ...lines,
];

const addEmptyLineAtEnd: Hook<unknown> = (lines) => [...lines, ''];

const defaultHooks = [labelAsGenerated, addEmptyLineAtEnd];

const processDatabase = async ({
  connection,
  preDeleteModelFolder = false,
  customTypeMap = {},

  modelHooks = [],
  modelNominator = nameIdentity,
  propertyNominator = (propertyName) =>
    propertyName.indexOf(' ') !== -1 ? `'${propertyName}'` : propertyName,
  initializerNominator = (modelName) => `${modelName}Initializer`,
  idNominator = (modelName) => `${modelName}Id`,

  makeIdType = (innerType, modelName) =>
    `${innerType} & { " __flavor"?: '${modelName}' };`,

  typeHooks = [],
  typeNominator = nameIdentity,

  fileNominator = identity,

  resolveViews = false,

  schemas,

  ...unknownProps
}: Config) => {
  if (!isEmpty(unknownProps)) {
    logger.warn(
      `Unknown configuration properties: ${Object.keys(unknownProps).join(
        ', '
      )}`
    );
  }

  const typeMap = { ...defaultTypeMap, ...customTypeMap };
  /** @type {import('./Config').Nominators} */
  const nominators = {
    modelNominator,
    propertyNominator,
    initializerNominator,
    idNominator,
    typeNominator,
    fileNominator,
  };
  const modelProcessChain = [...defaultHooks, ...modelHooks];
  const typeProcessChain = [...defaultHooks, ...typeHooks];

  if (typeof connection === 'string') {
    logger.log(`Connecting to ${chalk.greenBright(connection)}`);
  } else {
    logger.log(
      `Connecting to ${chalk.greenBright(connection.database)} on ${
        connection.host
      }`
    );
  }

  const schemaFolderMap = map(
    (s: SchemaConfig) => path.resolve(s.modelFolder),
    indexBy((s) => s.name, schemas)
  ) as Record<string, string>;

  for (const schemaConfig of schemas) {
    if (preDeleteModelFolder) {
      logger.log(` - Clearing old files in ${schemaConfig.modelFolder}`);
      await rmfr(schemaConfig.modelFolder, { glob: true });
    }
    if (!fs.existsSync(schemaConfig.modelFolder)) {
      fs.mkdirSync(schemaConfig.modelFolder, { recursive: true });
    }

    const schema = await extractSchema(
      schemaConfig.name,
      connection,
      schemaConfig.resolveViews !== undefined
        ? schemaConfig.resolveViews
        : resolveViews
    );

    await processSchema(
      schemaConfig,
      schema,
      typeMap,
      nominators,
      modelProcessChain,
      typeProcessChain,
      schemaFolderMap,
      makeIdType
    );
  }
};

export default processDatabase;
