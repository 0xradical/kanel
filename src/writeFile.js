import fs from 'fs';
import os from 'os';
import path from 'path';

import { logger } from './logger';

/**
 * @param {{ fullPath: string, lines: string[] }} p0
 */
const writeFile = ({ fullPath, lines }) => {
  const relativePath = path.relative(process.cwd(), fullPath);
  logger.log(` - ${relativePath}`);

  const content = lines.join(os.EOL);
  fs.writeFileSync(fullPath, content, 'utf-8');
};

export default writeFile;
