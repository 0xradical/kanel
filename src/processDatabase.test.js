import dvdRental from './mocks/dvdrental.json';
import processDatabase from './processDatabase';
import writeFile from './writeFile';

jest.mock('extract-pg-schema', () => ({
  extractSchema: () => dvdRental,
}));

jest.mock('./writeFile');
jest.mock('fs');
jest.mock('path', () => {
  const originalModule = jest.requireActual('path');
  return {
    __esModule: false,
    ...originalModule,
    resolve: (...args) => args.join('/'),
  };
});

describe('processDatabase', () => {
  it('should process the dvd rental example according to the snapshot', async () => {
    await processDatabase({
      connection: {
        host: 'localhost',
        user: 'postgres',
        password: 'postgres',
        database: 'dvdrental',
        charset: 'utf8',
        port: 54321,
      },

      preDeleteModelFolder: false,

      customTypeMap: {
        // There is no such package, this is just an example. See Film.ts to see the result.
        tsvector: {
          name: 'TsVector',
          module: 'ts-vector',
          absoluteImport: true,
          defaultImport: true,
        },
        bpchar: 'string',
      },

      resolveViews: true,

      schemas: [
        {
          name: 'public',
          modelFolder: '/models',
          ignore: ['film_list', 'staff'],
        },
      ],
    });

    // @ts-ignore
    const results = writeFile.mock.calls;
    expect(results).toMatchSnapshot();
  });
});
