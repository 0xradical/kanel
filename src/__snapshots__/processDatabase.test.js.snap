// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`processDatabase should process the dvd rental example according to the snapshot 1`] = `
Array [
  Array [
    Object {
      "fullPath": "/models/mpaa_rating.ts",
      "lines": Array [
        "// @generated",
        "// Automatically generated. Don't change this file manually.",
        "",
        "type mpaa_rating = 'G' | 'PG' | 'PG-13' | 'R' | 'NC-17';",
        "export default mpaa_rating;",
        "",
      ],
    },
  ],
  Array [
    Object {
      "fullPath": "/models/actor.ts",
      "lines": Array [
        "// @generated",
        "// Automatically generated. Don't change this file manually.",
        "",
        "export type actorId = number & { \\" __flavor\\"?: 'actor' };",
        "",
        "export default interface actor {",
        "  /** Primary key. Index: actor_pkey */",
        "  actor_id: actorId;",
        "",
        "  first_name: string;",
        "",
        "  /** Index: idx_actor_last_name */",
        "  last_name: string;",
        "",
        "  last_update: Date;",
        "}",
        "",
        "export interface actorInitializer {",
        "  /**",
        "   * Default value: nextval('actor_actor_id_seq'::regclass)",
        "   * Primary key. Index: actor_pkey",
        "   */",
        "  actor_id?: actorId;",
        "",
        "  first_name: string;",
        "",
        "  /** Index: idx_actor_last_name */",
        "  last_name: string;",
        "",
        "  /** Default value: now() */",
        "  last_update?: Date;",
        "}",
        "",
      ],
    },
  ],
  Array [
    Object {
      "fullPath": "/models/address.ts",
      "lines": Array [
        "// @generated",
        "// Automatically generated. Don't change this file manually.",
        "",
        "import { cityId } from './city';",
        "",
        "export type addressId = number & { \\" __flavor\\"?: 'address' };",
        "",
        "export default interface address {",
        "  /** Primary key. Index: address_pkey */",
        "  address_id: addressId;",
        "",
        "  address: string;",
        "",
        "  address2: string | null;",
        "",
        "  district: string;",
        "",
        "  /** Index: idx_fk_city_id */",
        "  city_id: cityId;",
        "",
        "  postal_code: string | null;",
        "",
        "  phone: string;",
        "",
        "  last_update: Date;",
        "}",
        "",
        "export interface addressInitializer {",
        "  /**",
        "   * Default value: nextval('address_address_id_seq'::regclass)",
        "   * Primary key. Index: address_pkey",
        "   */",
        "  address_id?: addressId;",
        "",
        "  address: string;",
        "",
        "  address2?: string | null;",
        "",
        "  district: string;",
        "",
        "  /** Index: idx_fk_city_id */",
        "  city_id: cityId;",
        "",
        "  postal_code?: string | null;",
        "",
        "  phone: string;",
        "",
        "  /** Default value: now() */",
        "  last_update?: Date;",
        "}",
        "",
      ],
    },
  ],
  Array [
    Object {
      "fullPath": "/models/category.ts",
      "lines": Array [
        "// @generated",
        "// Automatically generated. Don't change this file manually.",
        "",
        "export type categoryId = number & { \\" __flavor\\"?: 'category' };",
        "",
        "export default interface category {",
        "  /** Primary key. Index: category_pkey */",
        "  category_id: categoryId;",
        "",
        "  name: string;",
        "",
        "  last_update: Date;",
        "}",
        "",
        "export interface categoryInitializer {",
        "  /**",
        "   * Default value: nextval('category_category_id_seq'::regclass)",
        "   * Primary key. Index: category_pkey",
        "   */",
        "  category_id?: categoryId;",
        "",
        "  name: string;",
        "",
        "  /** Default value: now() */",
        "  last_update?: Date;",
        "}",
        "",
      ],
    },
  ],
  Array [
    Object {
      "fullPath": "/models/city.ts",
      "lines": Array [
        "// @generated",
        "// Automatically generated. Don't change this file manually.",
        "",
        "import { countryId } from './country';",
        "",
        "export type cityId = number & { \\" __flavor\\"?: 'city' };",
        "",
        "export default interface city {",
        "  /** Primary key. Index: city_pkey */",
        "  city_id: cityId;",
        "",
        "  city: string;",
        "",
        "  /** Index: idx_fk_country_id */",
        "  country_id: countryId;",
        "",
        "  last_update: Date;",
        "}",
        "",
        "export interface cityInitializer {",
        "  /**",
        "   * Default value: nextval('city_city_id_seq'::regclass)",
        "   * Primary key. Index: city_pkey",
        "   */",
        "  city_id?: cityId;",
        "",
        "  city: string;",
        "",
        "  /** Index: idx_fk_country_id */",
        "  country_id: countryId;",
        "",
        "  /** Default value: now() */",
        "  last_update?: Date;",
        "}",
        "",
      ],
    },
  ],
  Array [
    Object {
      "fullPath": "/models/country.ts",
      "lines": Array [
        "// @generated",
        "// Automatically generated. Don't change this file manually.",
        "",
        "export type countryId = number & { \\" __flavor\\"?: 'country' };",
        "",
        "export default interface country {",
        "  /** Primary key. Index: country_pkey */",
        "  country_id: countryId;",
        "",
        "  country: string;",
        "",
        "  last_update: Date;",
        "}",
        "",
        "export interface countryInitializer {",
        "  /**",
        "   * Default value: nextval('country_country_id_seq'::regclass)",
        "   * Primary key. Index: country_pkey",
        "   */",
        "  country_id?: countryId;",
        "",
        "  country: string;",
        "",
        "  /** Default value: now() */",
        "  last_update?: Date;",
        "}",
        "",
      ],
    },
  ],
  Array [
    Object {
      "fullPath": "/models/customer.ts",
      "lines": Array [
        "// @generated",
        "// Automatically generated. Don't change this file manually.",
        "",
        "import { addressId } from './address';",
        "",
        "export type customerId = number & { \\" __flavor\\"?: 'customer' };",
        "",
        "export default interface customer {",
        "  /** Primary key. Index: customer_pkey */",
        "  customer_id: customerId;",
        "",
        "  /** Index: idx_fk_store_id */",
        "  store_id: number;",
        "",
        "  first_name: string;",
        "",
        "  /** Index: idx_last_name */",
        "  last_name: string;",
        "",
        "  email: string | null;",
        "",
        "  /** Index: idx_fk_address_id */",
        "  address_id: addressId;",
        "",
        "  activebool: boolean;",
        "",
        "  create_date: Date;",
        "",
        "  last_update: Date | null;",
        "",
        "  active: number | null;",
        "}",
        "",
        "export interface customerInitializer {",
        "  /**",
        "   * Default value: nextval('customer_customer_id_seq'::regclass)",
        "   * Primary key. Index: customer_pkey",
        "   */",
        "  customer_id?: customerId;",
        "",
        "  /** Index: idx_fk_store_id */",
        "  store_id: number;",
        "",
        "  first_name: string;",
        "",
        "  /** Index: idx_last_name */",
        "  last_name: string;",
        "",
        "  email?: string | null;",
        "",
        "  /** Index: idx_fk_address_id */",
        "  address_id: addressId;",
        "",
        "  /** Default value: true */",
        "  activebool?: boolean;",
        "",
        "  /** Default value: ('now'::text)::date */",
        "  create_date?: Date;",
        "",
        "  /** Default value: now() */",
        "  last_update?: Date | null;",
        "",
        "  active?: number | null;",
        "}",
        "",
      ],
    },
  ],
  Array [
    Object {
      "fullPath": "/models/film.ts",
      "lines": Array [
        "// @generated",
        "// Automatically generated. Don't change this file manually.",
        "",
        "import { languageId } from './language';",
        "import mpaa_rating from './mpaa_rating';",
        "import TsVector from 'ts-vector';",
        "",
        "export type filmId = number & { \\" __flavor\\"?: 'film' };",
        "",
        "export default interface film {",
        "  /** Primary key. Index: film_pkey */",
        "  film_id: filmId;",
        "",
        "  /** Index: idx_title */",
        "  title: string;",
        "",
        "  description: string | null;",
        "",
        "  release_year: number | null;",
        "",
        "  /** Index: idx_fk_language_id */",
        "  language_id: languageId;",
        "",
        "  rental_duration: number;",
        "",
        "  rental_rate: string;",
        "",
        "  length: number | null;",
        "",
        "  replacement_cost: string;",
        "",
        "  rating: mpaa_rating | null;",
        "",
        "  last_update: Date;",
        "",
        "  special_features: text[] | null;",
        "",
        "  /** Index: film_fulltext_idx */",
        "  fulltext: TsVector;",
        "}",
        "",
        "export interface filmInitializer {",
        "  /**",
        "   * Default value: nextval('film_film_id_seq'::regclass)",
        "   * Primary key. Index: film_pkey",
        "   */",
        "  film_id?: filmId;",
        "",
        "  /** Index: idx_title */",
        "  title: string;",
        "",
        "  description?: string | null;",
        "",
        "  release_year?: number | null;",
        "",
        "  /** Index: idx_fk_language_id */",
        "  language_id: languageId;",
        "",
        "  /** Default value: 3 */",
        "  rental_duration?: number;",
        "",
        "  /** Default value: 4.99 */",
        "  rental_rate?: string;",
        "",
        "  length?: number | null;",
        "",
        "  /** Default value: 19.99 */",
        "  replacement_cost?: string;",
        "",
        "  /** Default value: 'G'::mpaa_rating */",
        "  rating?: mpaa_rating | null;",
        "",
        "  /** Default value: now() */",
        "  last_update?: Date;",
        "",
        "  special_features?: text[] | null;",
        "",
        "  /** Index: film_fulltext_idx */",
        "  fulltext: TsVector;",
        "}",
        "",
      ],
    },
  ],
  Array [
    Object {
      "fullPath": "/models/film_actor.ts",
      "lines": Array [
        "// @generated",
        "// Automatically generated. Don't change this file manually.",
        "",
        "import { actorId } from './actor';",
        "import { filmId } from './film';",
        "",
        "export default interface film_actor {",
        "  /** Primary key. Index: film_actor_pkey */",
        "  actor_id: actorId;",
        "",
        "  /**",
        "   * Primary key. Index: film_actor_pkey",
        "   * Index: idx_fk_film_id",
        "   */",
        "  film_id: filmId;",
        "",
        "  last_update: Date;",
        "}",
        "",
        "export interface film_actorInitializer {",
        "  /** Primary key. Index: film_actor_pkey */",
        "  actor_id: actorId;",
        "",
        "  /**",
        "   * Primary key. Index: film_actor_pkey",
        "   * Index: idx_fk_film_id",
        "   */",
        "  film_id: filmId;",
        "",
        "  /** Default value: now() */",
        "  last_update?: Date;",
        "}",
        "",
      ],
    },
  ],
  Array [
    Object {
      "fullPath": "/models/film_category.ts",
      "lines": Array [
        "// @generated",
        "// Automatically generated. Don't change this file manually.",
        "",
        "import { filmId } from './film';",
        "import { categoryId } from './category';",
        "",
        "export default interface film_category {",
        "  /** Primary key. Index: film_category_pkey */",
        "  film_id: filmId;",
        "",
        "  /** Primary key. Index: film_category_pkey */",
        "  category_id: categoryId;",
        "",
        "  last_update: Date;",
        "}",
        "",
        "export interface film_categoryInitializer {",
        "  /** Primary key. Index: film_category_pkey */",
        "  film_id: filmId;",
        "",
        "  /** Primary key. Index: film_category_pkey */",
        "  category_id: categoryId;",
        "",
        "  /** Default value: now() */",
        "  last_update?: Date;",
        "}",
        "",
      ],
    },
  ],
  Array [
    Object {
      "fullPath": "/models/inventory.ts",
      "lines": Array [
        "// @generated",
        "// Automatically generated. Don't change this file manually.",
        "",
        "import { filmId } from './film';",
        "",
        "export type inventoryId = number & { \\" __flavor\\"?: 'inventory' };",
        "",
        "export default interface inventory {",
        "  /** Primary key. Index: inventory_pkey */",
        "  inventory_id: inventoryId;",
        "",
        "  /** Index: idx_store_id_film_id */",
        "  film_id: filmId;",
        "",
        "  /** Index: idx_store_id_film_id */",
        "  store_id: number;",
        "",
        "  last_update: Date;",
        "}",
        "",
        "export interface inventoryInitializer {",
        "  /**",
        "   * Default value: nextval('inventory_inventory_id_seq'::regclass)",
        "   * Primary key. Index: inventory_pkey",
        "   */",
        "  inventory_id?: inventoryId;",
        "",
        "  /** Index: idx_store_id_film_id */",
        "  film_id: filmId;",
        "",
        "  /** Index: idx_store_id_film_id */",
        "  store_id: number;",
        "",
        "  /** Default value: now() */",
        "  last_update?: Date;",
        "}",
        "",
      ],
    },
  ],
  Array [
    Object {
      "fullPath": "/models/language.ts",
      "lines": Array [
        "// @generated",
        "// Automatically generated. Don't change this file manually.",
        "",
        "export type languageId = number & { \\" __flavor\\"?: 'language' };",
        "",
        "export default interface language {",
        "  /** Primary key. Index: language_pkey */",
        "  language_id: languageId;",
        "",
        "  name: string;",
        "",
        "  last_update: Date;",
        "}",
        "",
        "export interface languageInitializer {",
        "  /**",
        "   * Default value: nextval('language_language_id_seq'::regclass)",
        "   * Primary key. Index: language_pkey",
        "   */",
        "  language_id?: languageId;",
        "",
        "  name: string;",
        "",
        "  /** Default value: now() */",
        "  last_update?: Date;",
        "}",
        "",
      ],
    },
  ],
  Array [
    Object {
      "fullPath": "/models/payment.ts",
      "lines": Array [
        "// @generated",
        "// Automatically generated. Don't change this file manually.",
        "",
        "import { customerId } from './customer';",
        "import { staffId } from './staff';",
        "import { rentalId } from './rental';",
        "",
        "export type paymentId = number & { \\" __flavor\\"?: 'payment' };",
        "",
        "export default interface payment {",
        "  /** Primary key. Index: payment_pkey */",
        "  payment_id: paymentId;",
        "",
        "  /** Index: idx_fk_customer_id */",
        "  customer_id: customerId;",
        "",
        "  /** Index: idx_fk_staff_id */",
        "  staff_id: staffId;",
        "",
        "  /** Index: idx_fk_rental_id */",
        "  rental_id: rentalId;",
        "",
        "  amount: string;",
        "",
        "  payment_date: Date;",
        "}",
        "",
        "export interface paymentInitializer {",
        "  /**",
        "   * Default value: nextval('payment_payment_id_seq'::regclass)",
        "   * Primary key. Index: payment_pkey",
        "   */",
        "  payment_id?: paymentId;",
        "",
        "  /** Index: idx_fk_customer_id */",
        "  customer_id: customerId;",
        "",
        "  /** Index: idx_fk_staff_id */",
        "  staff_id: staffId;",
        "",
        "  /** Index: idx_fk_rental_id */",
        "  rental_id: rentalId;",
        "",
        "  amount: string;",
        "",
        "  payment_date: Date;",
        "}",
        "",
      ],
    },
  ],
  Array [
    Object {
      "fullPath": "/models/rental.ts",
      "lines": Array [
        "// @generated",
        "// Automatically generated. Don't change this file manually.",
        "",
        "import { inventoryId } from './inventory';",
        "import { customerId } from './customer';",
        "import { staffId } from './staff';",
        "",
        "export type rentalId = number & { \\" __flavor\\"?: 'rental' };",
        "",
        "export default interface rental {",
        "  /** Primary key. Index: rental_pkey */",
        "  rental_id: rentalId;",
        "",
        "  /** Index: idx_unq_rental_rental_date_inventory_id_customer_id */",
        "  rental_date: Date;",
        "",
        "  /**",
        "   * Index: idx_fk_inventory_id",
        "   * Index: idx_unq_rental_rental_date_inventory_id_customer_id",
        "   */",
        "  inventory_id: inventoryId;",
        "",
        "  /** Index: idx_unq_rental_rental_date_inventory_id_customer_id */",
        "  customer_id: customerId;",
        "",
        "  return_date: Date | null;",
        "",
        "  staff_id: staffId;",
        "",
        "  last_update: Date;",
        "}",
        "",
        "export interface rentalInitializer {",
        "  /**",
        "   * Default value: nextval('rental_rental_id_seq'::regclass)",
        "   * Primary key. Index: rental_pkey",
        "   */",
        "  rental_id?: rentalId;",
        "",
        "  /** Index: idx_unq_rental_rental_date_inventory_id_customer_id */",
        "  rental_date: Date;",
        "",
        "  /**",
        "   * Index: idx_fk_inventory_id",
        "   * Index: idx_unq_rental_rental_date_inventory_id_customer_id",
        "   */",
        "  inventory_id: inventoryId;",
        "",
        "  /** Index: idx_unq_rental_rental_date_inventory_id_customer_id */",
        "  customer_id: customerId;",
        "",
        "  return_date?: Date | null;",
        "",
        "  staff_id: staffId;",
        "",
        "  /** Default value: now() */",
        "  last_update?: Date;",
        "}",
        "",
      ],
    },
  ],
  Array [
    Object {
      "fullPath": "/models/store.ts",
      "lines": Array [
        "// @generated",
        "// Automatically generated. Don't change this file manually.",
        "",
        "import { staffId } from './staff';",
        "import { addressId } from './address';",
        "",
        "export type storeId = number & { \\" __flavor\\"?: 'store' };",
        "",
        "export default interface store {",
        "  /** Primary key. Index: store_pkey */",
        "  store_id: storeId;",
        "",
        "  /** Index: idx_unq_manager_staff_id */",
        "  manager_staff_id: staffId;",
        "",
        "  address_id: addressId;",
        "",
        "  last_update: Date;",
        "}",
        "",
        "export interface storeInitializer {",
        "  /**",
        "   * Default value: nextval('store_store_id_seq'::regclass)",
        "   * Primary key. Index: store_pkey",
        "   */",
        "  store_id?: storeId;",
        "",
        "  /** Index: idx_unq_manager_staff_id */",
        "  manager_staff_id: staffId;",
        "",
        "  address_id: addressId;",
        "",
        "  /** Default value: now() */",
        "  last_update?: Date;",
        "}",
        "",
      ],
    },
  ],
  Array [
    Object {
      "fullPath": "/models/actor_info.ts",
      "lines": Array [
        "// @generated",
        "// Automatically generated. Don't change this file manually.",
        "",
        "export default interface actor_info {",
        "  actor_id: number | null;",
        "",
        "  first_name: string | null;",
        "",
        "  last_name: string | null;",
        "",
        "  film_info: string | null;",
        "}",
        "",
      ],
    },
  ],
  Array [
    Object {
      "fullPath": "/models/customer_list.ts",
      "lines": Array [
        "// @generated",
        "// Automatically generated. Don't change this file manually.",
        "",
        "export default interface customer_list {",
        "  id: number | null;",
        "",
        "  name: string | null;",
        "",
        "  address: string | null;",
        "",
        "  'zip code': string | null;",
        "",
        "  phone: string | null;",
        "",
        "  city: string | null;",
        "",
        "  country: string | null;",
        "",
        "  notes: string | null;",
        "",
        "  sid: number | null;",
        "}",
        "",
      ],
    },
  ],
  Array [
    Object {
      "fullPath": "/models/nicer_but_slower_film_list.ts",
      "lines": Array [
        "// @generated",
        "// Automatically generated. Don't change this file manually.",
        "",
        "import mpaa_rating from './mpaa_rating';",
        "",
        "export default interface nicer_but_slower_film_list {",
        "  fid: number | null;",
        "",
        "  title: string | null;",
        "",
        "  description: string | null;",
        "",
        "  category: string | null;",
        "",
        "  price: string | null;",
        "",
        "  length: number | null;",
        "",
        "  rating: mpaa_rating | null;",
        "",
        "  actors: string | null;",
        "}",
        "",
      ],
    },
  ],
  Array [
    Object {
      "fullPath": "/models/sales_by_film_category.ts",
      "lines": Array [
        "// @generated",
        "// Automatically generated. Don't change this file manually.",
        "",
        "export default interface sales_by_film_category {",
        "  category: string | null;",
        "",
        "  total_sales: string | null;",
        "}",
        "",
      ],
    },
  ],
  Array [
    Object {
      "fullPath": "/models/sales_by_store.ts",
      "lines": Array [
        "// @generated",
        "// Automatically generated. Don't change this file manually.",
        "",
        "export default interface sales_by_store {",
        "  store: string | null;",
        "",
        "  manager: string | null;",
        "",
        "  total_sales: string | null;",
        "}",
        "",
      ],
    },
  ],
  Array [
    Object {
      "fullPath": "/models/staff_list.ts",
      "lines": Array [
        "// @generated",
        "// Automatically generated. Don't change this file manually.",
        "",
        "export default interface staff_list {",
        "  id: number | null;",
        "",
        "  name: string | null;",
        "",
        "  address: string | null;",
        "",
        "  'zip code': string | null;",
        "",
        "  phone: string | null;",
        "",
        "  city: string | null;",
        "",
        "  country: string | null;",
        "",
        "  sid: number | null;",
        "}",
        "",
      ],
    },
  ],
  Array [
    Object {
      "fullPath": "/models/index.ts",
      "lines": Array [
        "// @generated",
        "// Automatically generated. Don't change this file manually.",
        "",
        "import actor, { actorInitializer, actorId } from './actor';",
        "import address, { addressInitializer, addressId } from './address';",
        "import category, { categoryInitializer, categoryId } from './category';",
        "import city, { cityInitializer, cityId } from './city';",
        "import country, { countryInitializer, countryId } from './country';",
        "import customer, { customerInitializer, customerId } from './customer';",
        "import film, { filmInitializer, filmId } from './film';",
        "import film_actor, { film_actorInitializer } from './film_actor';",
        "import film_category, { film_categoryInitializer } from './film_category';",
        "import inventory, { inventoryInitializer, inventoryId } from './inventory';",
        "import language, { languageInitializer, languageId } from './language';",
        "import payment, { paymentInitializer, paymentId } from './payment';",
        "import rental, { rentalInitializer, rentalId } from './rental';",
        "import store, { storeInitializer, storeId } from './store';",
        "import actor_info from './actor_info';",
        "import customer_list from './customer_list';",
        "import nicer_but_slower_film_list from './nicer_but_slower_film_list';",
        "import sales_by_film_category from './sales_by_film_category';",
        "import sales_by_store from './sales_by_store';",
        "import staff_list from './staff_list';",
        "import mpaa_rating from './mpaa_rating';",
        "",
        "type Model =",
        "  | actor",
        "  | address",
        "  | category",
        "  | city",
        "  | country",
        "  | customer",
        "  | film",
        "  | film_actor",
        "  | film_category",
        "  | inventory",
        "  | language",
        "  | payment",
        "  | rental",
        "  | store",
        "  | actor_info",
        "  | customer_list",
        "  | nicer_but_slower_film_list",
        "  | sales_by_film_category",
        "  | sales_by_store",
        "  | staff_list",
        "",
        "interface ModelTypeMap {",
        "  'actor': actor;",
        "  'address': address;",
        "  'category': category;",
        "  'city': city;",
        "  'country': country;",
        "  'customer': customer;",
        "  'film': film;",
        "  'film_actor': film_actor;",
        "  'film_category': film_category;",
        "  'inventory': inventory;",
        "  'language': language;",
        "  'payment': payment;",
        "  'rental': rental;",
        "  'store': store;",
        "  'actor_info': actor_info;",
        "  'customer_list': customer_list;",
        "  'nicer_but_slower_film_list': nicer_but_slower_film_list;",
        "  'sales_by_film_category': sales_by_film_category;",
        "  'sales_by_store': sales_by_store;",
        "  'staff_list': staff_list;",
        "}",
        "",
        "type ModelId =",
        "  | actorId",
        "  | addressId",
        "  | categoryId",
        "  | cityId",
        "  | countryId",
        "  | customerId",
        "  | filmId",
        "  | inventoryId",
        "  | languageId",
        "  | paymentId",
        "  | rentalId",
        "  | storeId",
        "",
        "interface ModelIdTypeMap {",
        "  'actor': actorId;",
        "  'address': addressId;",
        "  'category': categoryId;",
        "  'city': cityId;",
        "  'country': countryId;",
        "  'customer': customerId;",
        "  'film': filmId;",
        "  'inventory': inventoryId;",
        "  'language': languageId;",
        "  'payment': paymentId;",
        "  'rental': rentalId;",
        "  'store': storeId;",
        "}",
        "",
        "type Initializer =",
        "  | actorInitializer",
        "  | addressInitializer",
        "  | categoryInitializer",
        "  | cityInitializer",
        "  | countryInitializer",
        "  | customerInitializer",
        "  | filmInitializer",
        "  | film_actorInitializer",
        "  | film_categoryInitializer",
        "  | inventoryInitializer",
        "  | languageInitializer",
        "  | paymentInitializer",
        "  | rentalInitializer",
        "  | storeInitializer",
        "",
        "interface InitializerTypeMap {",
        "  'actor': actorInitializer;",
        "  'address': addressInitializer;",
        "  'category': categoryInitializer;",
        "  'city': cityInitializer;",
        "  'country': countryInitializer;",
        "  'customer': customerInitializer;",
        "  'film': filmInitializer;",
        "  'film_actor': film_actorInitializer;",
        "  'film_category': film_categoryInitializer;",
        "  'inventory': inventoryInitializer;",
        "  'language': languageInitializer;",
        "  'payment': paymentInitializer;",
        "  'rental': rentalInitializer;",
        "  'store': storeInitializer;",
        "}",
        "",
        "export type {",
        "  actor, actorInitializer, actorId,",
        "  address, addressInitializer, addressId,",
        "  category, categoryInitializer, categoryId,",
        "  city, cityInitializer, cityId,",
        "  country, countryInitializer, countryId,",
        "  customer, customerInitializer, customerId,",
        "  film, filmInitializer, filmId,",
        "  film_actor, film_actorInitializer,",
        "  film_category, film_categoryInitializer,",
        "  inventory, inventoryInitializer, inventoryId,",
        "  language, languageInitializer, languageId,",
        "  payment, paymentInitializer, paymentId,",
        "  rental, rentalInitializer, rentalId,",
        "  store, storeInitializer, storeId,",
        "  actor_info,",
        "  customer_list,",
        "  nicer_but_slower_film_list,",
        "  sales_by_film_category,",
        "  sales_by_store,",
        "  staff_list,",
        "  mpaa_rating,",
        "",
        "  Model,",
        "  ModelTypeMap,",
        "  ModelId,",
        "  ModelIdTypeMap,",
        "  Initializer,",
        "  InitializerTypeMap",
        "};",
        "",
      ],
    },
  ],
]
`;
